kernel/boot_info/src/multiboot2.rs

```diff
@@ -0,0 +1,176 @@
+        let start = PhysicalAddress::new(
+            self.elf_sections()?
+                .into_iter()
+                .filter(|section| section.is_allocated())
+                .map(|section| section.start_address())
+                .min()
+                .ok_or("couldn't find kernel start address")? as usize,
+        )
+        .ok_or("kernel physical start address was invalid")?;
+
+        let virtual_end = VirtualAddress::new(
+            self.elf_sections()?
+                .into_iter()
+                .filter(|s| s.is_allocated())
+                .map(|s| s.end_address())
+                .max()
+                .ok_or("couldn't find kernel end address")? as usize,
+        )
+        .ok_or("kernel virtual end address was invalid")?;
+        let physical_end = PhysicalAddress::new(virtual_end.value() - KERNEL_OFFSET)
+            .ok_or("kernel physical end address was invalid")?;
+
+        Ok(start..physical_end)
+    }
+
+    fn bootloader_info_memory_range(&self) -> Result<Range<PhysicalAddress>, &'static str> {
+        let start = PhysicalAddress::new(self.start_address() - KERNEL_OFFSET)
+            .ok_or("invalid bootloader info start address")?;
+        let end = PhysicalAddress::new(self.end_address() - KERNEL_OFFSET)
+            .ok_or("invalid bootloader info end address")?;
+        Ok(start..end)
+    }
+
+    fn modules_memory_range(&self) -> Result<Range<PhysicalAddress>, &'static str> {
+        let mut min = usize::MAX;
+        let mut max = 0;
+
+        for module in self.module_tags() {
+            min = cmp::min(min, module.start_address() as usize);
+            max = cmp::max(max, module.end_address() as usize);
+        }
+
+        Ok(PhysicalAddress::new_canonical(min)..PhysicalAddress::new_canonical(max))
+    }
+
+    fn memory_regions(&self) -> Result<Self::MemoryRegions<'_>, &'static str> {
+        Ok(MemoryRegions {
+            inner: self
+                .memory_map_tag()
+                .ok_or("no memory map tag")?
+                .memory_areas(),
+        })
+    }
+
+    fn elf_sections(&self) -> Result<Self::ElfSections<'static>, &'static str> {
+        Ok(self
+            .elf_sections_tag()
+            .ok_or("no elf sections tag")?
+            .sections())
+    }
+
+    fn modules(&self) -> Self::Modules<'_> {
+        self.module_tags()
+    }
+
+    fn rsdp(&self) -> Option<PhysicalAddress> {
+        // FIXME: Implement. Depends on #734.
+        None
+    }
+
+    fn stack_size(&self) -> usize {
+        use crate::ElfSection;
+
+        self.elf_sections()
+            .expect("couldn't get elf sections")
+            .filter(|section| section.name() == ".stack")
+            .map(|section| {
+                let start = section.start();
+                let end = start + section.len();
+                (end - start).value()
+            })
+            .next()
+            .expect("no stack section")
```

Comment:
why does only this function panic? Might as well return a `Result` here too, just like for all the other fallible functions.